prow/cmd/phaino/local.go

```diff
@@ -0,0 +1,339 @@
+	var out string
+	fmt.Scanln(&out)
+	return realPath(out)
+}
+
+func volume(pod coreapi.PodSpec, name string) *coreapi.Volume {
+	for _, v := range pod.Volumes {
+		if v.Name == name {
+			return &v
+		}
+	}
+	return nil
+}
+
+func pathAlias(r prowapi.Refs) string {
+	if r.PathAlias == "" {
+		return fmt.Sprintf("github.com/%s/%s", r.Org, r.Repo)
+	}
+	return r.PathAlias
+}
+
+func readRepo(path string) (string, error) {
+	wd, err := workingDir()
+	if err != nil {
+		return "", fmt.Errorf("workingDir: %v", err)
+	}
+	def, err := findRepo(wd, path)
+	if err != nil {
+		logrus.WithError(err).WithField("repo", path).Warn("could not find repo")
+	}
+	fmt.Fprintf(os.Stderr, "local /path/to/%s", path)
+	if def != "" {
+		fmt.Fprintf(os.Stderr, " [%s]", def)
+	}
+	fmt.Fprint(os.Stderr, ": ")
+	var out string
+	fmt.Scanln(&out)
+	if out == "" {
+		out = def
+	}
+	return realPath(out)
+}
+
+func workingDir() (string, error) {
+	if wd := os.Getenv("BUILD_WORKING_DIRECTORY"); wd != "" {
+		return wd, nil // running via bazel run
+	}
+	return os.Getwd() // running outside bazel
+}
+
+// findRepo will attempt to find a repo in logical locations under path.
+//
+// It will first try to find foo/bar somewhere under $PWD or a $PWD dir.
+// AKA if $PWD is /go/src it will match /go/src/foo/bar, /go/foo/bar or /foo/bar
+// Next it will look for the basename somewhere under $PWD or a $PWD dir.
+// AKA if $PWD is /go/src it will match /go/src/bar, /go/bar or /bar
+// If both of these strategies fail it will return an error.
+func findRepo(wd, path string) (string, error) {
+	opwd, err := realPath(wd)
+	if err != nil {
+		return "", fmt.Errorf("wd not found: %v", err)
+	}
+	if strings.HasPrefix(path, "github.com/kubernetes/") {
+		path = strings.Replace(path, "github.com/kubernetes/", "k8s.io/", 1)
+	}
+
+	var old string
+	pwd := opwd
+	for old != pwd {
+		old = pwd
+		if strings.HasSuffix(pwd, "/"+path) {
+			return pwd, nil
+		}
+		pwd = filepath.Dir(pwd)
+	}
+	pwd = opwd
+	for old != pwd {
+		old = pwd
+		check := filepath.Join(pwd, path)
+		if info, err := os.Stat(check); err == nil && info.IsDir() {
+			return check, nil
+		}
+		pwd = filepath.Dir(pwd)
+	}
+
+	base := filepath.Base(path)
+	pwd = opwd
+	for old != pwd {
+		old = pwd
+		check := filepath.Join(pwd, base)
+		if info, err := os.Stat(check); err == nil && info.IsDir() {
+			return check, nil
+		}
+		pwd = filepath.Dir(pwd)
+	}
+	return "", errors.New("cannot find repo")
+}
+
+var baseArgs = []string{"docker", "run", "--rm=true"}
+
+func checkPrivilege(cont coreapi.Container, allow bool) (bool, error) {
+	if cont.SecurityContext == nil {
+		return false, nil
+	}
+	if cont.SecurityContext.Privileged == nil {
+		return false, nil
+	}
+	if !*cont.SecurityContext.Privileged {
+		return false, nil
+	}
+	fmt.Fprint(os.Stderr, "Privileged jobs are unsafe. Remove from local run? [yes]: ")
+	var out string
+	fmt.Scanln(&out)
+	if out == "" || out[0] == 'y' || out[0] == 'Y' {
```

Comment:
"yes" (or other case variations thereof?)

Possibly: require an understandable "no" to not remove, rather than an understandable "yes" to remove, because "no" is the unsafe choice.

Reply:
Done

(yes will match `out[0] == 'y'`, right?)

Reply:
(evidently I cannot read.)